pkg/jsonmerge/object.go

```diff
@@ -0,0 +1,73 @@
+package jsonmerge
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+
+	"dario.cat/mergo"
+)
+
+type objectMerger struct {
+	io.Writer
+	dst map[string]interface{}
+}
+
+// NewObjectMerger creates a Merger for JSON objects.
+func NewObjectMerger(w io.Writer) Merger {
+	return &objectMerger{
+		Writer: w,
+		dst:    make(map[string]interface{}),
+	}
+}
+
+func (merger *objectMerger) NewPage(r io.Reader, isLastPage bool) io.ReadCloser {
+	return &objectMergerPage{
+		merger: merger,
+		Reader: r,
+	}
+}
+
+func (merger *objectMerger) Close() error {
+	// Marshal to JSON and write to output.
+	buf, err := json.Marshal(merger.dst)
+	if err != nil {
+		return err
+	}
+
+	_, err = merger.Writer.Write(buf)
+	return err
+}
+
+type objectMergerPage struct {
+	merger *objectMerger
+
+	io.Reader
+	buffer bytes.Buffer
+}
+
+func (page *objectMergerPage) Read(p []byte) (int, error) {
+	// Read into a temporary buffer to be merged and written later.
+	p = make([]byte, len(p))
```

Comment:
Could this whole function be collapsed to:

```go
func (page *objectMergerPage) Read(p []byte) (int, error) {
	_, err := io.CopyN(&page.buffer, page.Reader, int64(len(p)))
	return 0, err
}
```

Reply:
This works, but I got to thinking: the original code is wrong, and it's not fixed by using `io.CopyN`. We may actually read less than `len(p)`.

That said, I supposed it doesn't matter given the docs:

> CopyN copies n bytes (or until an error) from src to dst. It returns the number of bytes copied and the earliest error encountered while copying. On return, written == n if and only if err == nil.

We'd get an error if `written != n`.